src/shims/native_lib/ffi.rs

```diff
@@ -0,0 +1,113 @@
+use libffi::low::CodePtr;
+use libffi::middle::{Arg as ArgPtr, Cif, Type as FfiType};
+
+/// Perform the actual FFI call.
+///
+/// SAFETY: The `FfiArg`s passed must have been correctly instantiated (i.e. their
+/// type layout must match the data they point to), and the safety invariants of
+/// the foreign function being called must be upheld (if any).
+pub unsafe fn call<'a, R: libffi::high::CType>(fun: CodePtr, args: Vec<FfiArg<'a>>) -> R {
+    let mut arg_tys = vec![];
+    let mut arg_ptrs = vec![];
+    for arg in args {
+        arg_tys.push(arg.ty);
+        arg_ptrs.push(arg.ptr)
+    }
+    let cif = Cif::new(arg_tys, R::reify().into_middle());
+    unsafe { cif.call(fun, &arg_ptrs) }
+}
+
+/// A wrapper type for `libffi::middle::Type` which also holds a pointer to the data.
+pub struct FfiArg<'a> {
+    /// The type layout information for the pointed-to data.
+    ty: FfiType,
+    /// A pointer to the data described in `ty`.
+    ptr: ArgPtr,
+    /// Lifetime of the actual pointed-to data.
+    _p: std::marker::PhantomData<&'a [u8]>,
+}
+
+impl<'a> FfiArg<'a> {
+    fn new(ty: FfiType, ptr: ArgPtr) -> Self {
+        Self { ty, ptr, _p: std::marker::PhantomData }
+    }
+}
+
+/// An owning form of `FfiArg`.
```

Comment:
The prefixes "C" and "Ffi" don't do a great job reflecting "owned" and borrowed.

So what about `OwnedArg` and `BorrowedArg`? The "downcast" function (also an odd name, what does it downcast?) can then be called `borrow`.

`CPrimitive` could be called `ScalarArg` as that seems to match what we use it for.

Also, please generally define types bottom-up and with increasing complexity/subtlety, to simplify reading the file in order:
- ScalarArg
- OwnedArg
- BorrowedArg